node/src/components/consensus/highway_core/state/panorama.rs

```diff
@@ -187,4 +187,29 @@ impl<C: Context> Panorama<C> {
         }
         Ok(())
     }
+
+    // Returns a bit mask that sets bit to 1 for every fault validator at equivocator's
+    // `ValidatorIndex`.
+    pub(crate) fn faults_bit_mask(&self) -> usize {
+        let mut bit_mask = 0;
+        for (idx, _) in self.enumerate().filter(|&(_, obs)| obs.is_faulty()) {
+            bit_mask |= 1 << idx.0;
+        }
+        bit_mask
+    }
+
+    /// Returns votes' hashes from this panorama that cite naively equivocator's vote.
+    pub(crate) fn need_endorsements(&self, state: &State<C>) -> Vec<&C::Hash> {
+        let local_faults = state.panorama.faults_bit_mask();
+        self.iter_correct_hashes()
+            .filter(|&v| {
+                let vote = state.vote(v);
+                let vote_faults = vote.panorama.faults_bit_mask();
+                // XOR will return 1 in places where we had seen equivocation locally
+                // but `vote_faults` did not (i.e. validator is still seen as correct).
+                // XOR != 0 means that `vote` must be endorsed b/c it cites _an_ equivocator.
+                local_faults ^ vote_faults != 0
```

Comment:
Since `self` is later than `vote.panorama`, everyone seen as faulty by `vote.panorama` is also seen as faulty by `self`, so this is really just about the case where `vote_faults` is `false` and `local_faults` is `true`. I think the bitmask is overkill, and maybe doesn't even improve performance at all. I'd just collect the validators seen as faulty by `self` (by index), and then check whether `vote_faults[idx].is_faulty()` for each of them.

Reply:
You're probably right :(

Reply:
Done.